Fix required opt validation bugs in CLI

ghubstan opened the pull request:
Many required CLI opt checks were incorrectly passing due to assignment of `""` default values.  Server side validation was helping cover up some of affects of this type of bug, but this change resolves the issue in the CLI side.

Credit to @BtcContributor for exposing the problem while beta testing the api.

Changes are listed below.

Unsets empty string default values on some the `createoffer` method's required option declarations in CreateOfferOptionParser, and adds a check for presence of required currency-code option.

- Removes default "" value from required opt direction.
- Removes default "" value from required opt currency-code.
- Removes default "" value from required opt amount.
- Removes default "" value from required opt min-amount.
- Removes default "" value from required opt fixed-price.
- Removes default "" value from required opt security-deposit.
- Check for required currency-code option.

Unsets empty string default values on some of the `takeoffer` method's required option declarations in TakeOfferOptionParser.

- Removes default "" value from required opt offer-id.
- Removes default "" value from required opt payment-account.

Other opt parser default values removed from:

- CancelOfferOptionParser#offer-id
- CreatePaymentAcctOptionParser#payment-account-form
- GetAddressBalanceOptionParser#address
- GetBTCMarketPriceOptionParser#currency-code
- GetOfferOptionParser#offer-id
- GetOffersOptionParser#direction
- GetOffersOptionParser#currency-code
- GetPaymentAcctFormOptionParser#payment-method-id
- GetTradeOptionParser#trade-id
- GetTransactionOptionParser#transaction-id
- RegisterDisputeAgentOptionParser#registration-key
- RegisterDisputeAgentOptionParser#dispute-agent-type
- RemoveWalletPasswordOptionParser#wallet-password
- SendBsqOptionParser#address
- SendBsqOptionParser#amount
- SendBtcOptionParser#address
- SendBtcOptionParser#amount
- SetTxFeeRateOptionParser#tx-fee-rate
- SetWalletPasswordOptionParser#wallet-password
- UnlockWalletOptionParser#wallet-password
- UnlockWalletOptionParser#timeout
- WithdrawFundsOptionParser#trade-id
- WithdrawFundsOptionParser#address

PR https://github.com/bisq-network/bisq/pull/5273 should be reviewed & merged before this one.

ghubstan committed:
Display buyer's cost in api's gettrade output

The offer volume is shown so traders know how much fiat
they are sending or receiving without having to call getoffer.

Changed the 'Fiat Sent' and 'Fiat Received' column headers
to show which fiat is being transfered, e.g., 'EUR Sent',
'EUR Received'.

Adjusted apitest's trade-simulation-utils.sh to the modified
gettrade output.

ghubstan committed:
Merge branch 'master' into 01-show-cost-in-trade-output

ghubstan committed:
Use the logger of the gRPC service throwing an exception

ghubstan committed:
Permit some gRPC excptions to be logged only as warning

A new handleExceptionAsWarning method logs warn(ex.msg) instead of
the full stack trace.

ghubstan committed:
Log 'trade not found' a warning instead of full stack trace

ghubstan committed:
Fix test bug

ghubstan committed:
Clean up call rate meter config file in test teardown

ghubstan committed:
Fix test bug

ghubstan committed:
Fix call rate metering interceptor bug

The gRPC interceptor was not using the correct method/ratemeter
map key, and failing to find a rate meter for the server call.

- Fix the rate meter key lookup bug.

- Disable most strict, default call rate meters in tests.
  Made an adjustment to the test harness' scaffold setup so an interceptor
  testing or disabling config file is always picked up by bob and alice
  daemons.

- Set arbitration daemon's registerDisputeAgent rate @ 10/second, so
  it does not interfere with the test harness.  (There is no pre-existing
  arb node appDataDir before that daemon starts.)

Note:  The test harness cannot install the custom rate metering file in
an arb daemon's appDataDir before it starts because there is no dao-setup
file for that node type.

TODO:  Adjust api simulation scripts to interceptor bug fix.

ghubstan committed:
Make test call rate = default call rate

ghubstan committed:
No need to wait, default+test call rate > 2x / second

ghubstan committed:
Remove unused import

ghubstan committed:
Adjust mainnet bats test to default rate meter interceptors

ghubstan committed:
Wait 3 secs after removing password (for wallet save)

ghubstan committed:
Fix getunusedbsqaddress test

ghubstan committed:
Improve interceptor's rate metering key definition and lookup

This change replaces the hard coded strings used as keys in
interceptor rate-metering lookup maps.

Now, the fullMethodName defined in each bisq.proto.grpc.* class'
io.grpc.MethodDescriptor is used, not a hard coded String.

For example, the rate metering lookup key for 'GetBalances',
in 'GrpcWalletsService', is the

   fullMethodName = SERVICE_NAME + '/' + "GetBalances",

   where SERVICE_NAME = "io.bisq.protobuffer.Wallets".

Also adjusted affected tests, and tidy'd up interceptor logging.

ghubstan committed:
Fix CLI number opt validation, improve server-not-up msg

- Fix tx-fee-rate opt validation bug.

- Tell user what option value is not a number.

- Append ", server may not be running" text to "io exception" exception msg.

ghubstan committed:
Fix tx-fee-rate formatting (and math) bug in cli/CurrencyFormat

Now prints '<integer> sats/byte' to console.

ghubstan committed:
Remove warning supression

ghubstan committed:
Add link to api-beta-test-guide.md

ghubstan committed:
Explain how to manually register test dispute agents

ghubstan committed:
Fix opt validation bugs in CLI

Unset empty string default values on some the createoffer method's
required option declarations in CreateOfferOptionParser, and adds a
check for presence of required currency-code option.

- Removes default "" value from required opt direction.
- Removes default "" value from required opt currency-code.
- Removes default "" value from required opt amount.
- Removes default "" value from required opt min-amount.
- Removes default "" value from required opt fixed-price.
- Removes default "" value from required opt security-deposit.
- Check for required currency-code option.

Unset empty string default values on some of the createoffer method's
required option declarations in TakeOfferOptionParser.

- Removes default "" value from required opt offer-id.
- Removes default "" value from required opt payment-account.

Other opt parser default values removed from:

- CancelOfferOptionParser#offer-id
- CreatePaymentAcctOptionParser#payment-account-form
- GetAddressBalanceOptionParser#address
- GetBTCMarketPriceOptionParser#currency-code
- GetOfferOptionParser#offer-id
- GetOffersOptionParser#direction
- GetOffersOptionParser#currency-code
- GetPaymentAcctFormOptionParser#payment-method-id
- GetTradeOptionParser#trade-id
- GetTransactionOptionParser#transaction-id
- RegisterDisputeAgentOptionParser#registration-key
- RegisterDisputeAgentOptionParser#dispute-agent-type
- RemoveWalletPasswordOptionParser#wallet-password
- SendBsqOptionParser#address
- SendBsqOptionParser#amount
- SendBtcOptionParser#address
- SendBtcOptionParser#amount
- SetTxFeeRateOptionParser#tx-fee-rate
- SetWalletPasswordOptionParser#wallet-password
- UnlockWalletOptionParser#wallet-password
- UnlockWalletOptionParser#timeout
- WithdrawFundsOptionParser#trade-id
- WithdrawFundsOptionParser#address

ghubstan committed:
Update cli getoffers smoke test to posix style opts

ghubstan committed:
Improve required-argument opt validation

Make sure non-empty opt values are present in the parsers' custom
validation, or jsimpleopt will throw exceptions with stylistically
inconsisent messages.

ghubstan committed:
Add jupiter test support to :cli subproject

ghubstan committed:
Add new cli option parser test

Does not check every combination of options for every method, but
we finally have an option parser test case to build on.

ghubstan reviewed (commented):
The last part of commit d01a7b7feec7167b66ec808529d2e63ffe248442 's message is incorrect.  Commit d01a7b7feec7167b66ec808529d2e63ffe248442 makes _empty_ opt values invalid, and throws an exception with the conventional `no <opt> specified` error message. Before this change, the CLI's opt parser accepted `--offer-id=` as a valid option, with an empty value.

The _next_ commit makes the option parser throw option parser exceptions due to  _missing (not empty)_ option values.  The next commit will result in a stylistically consistent error message printed to the console when `--offer-id` is present without an `=` delimiter (as if it were a no-arg option).

ghubstan committed:
Handle require-arg options missing the = sign

This change ensure a stylistically consistent error message is printed
to the CLI console if an option is present, but without the '=' sign.

sqrrm commented on cli/src/test/java/bisq/cli/opt/OptionParsersTest.java:
```diff
@@ -0,0 +1,160 @@
+                "--" + OPT_AMOUNT + "=" + "0.125",
+                "--" + OPT_MKT_PRICE_MARGIN + "=" + "0.0",
+                "--" + OPT_SECURITY_DEPOSIT + "=" + "25.0"
+        };
```
Is there an assert missing here?

ghubstan replied:
Added in commit 70da6d19fa694c5ae11212645fdfa4ea47529854.

ghubstan replied:
In next PR ([5284](https://github.com/bisq-network/bisq/pull/5284)) more asserts were added to check valid, parsed opt values in commit 389abf4.

```
	assertEquals("abc-payment-acct-id-123", parser.getPaymentAccountId());
	assertEquals("BUY", parser.getDirection());
	assertEquals("EUR", parser.getCurrencyCode());
	assertEquals("0.125", parser.getAmount());
	assertEquals("0.0", parser.getMktPriceMargin());
	assertEquals("25.0", parser.getSecurityDeposit());
```

ghubstan committed:
Parse args in opts test, no exception = pass

Resolves PR review issue
https://github.com/bisq-network/bisq/pull/5274/files/304781ce92c13df187b26d1ac2ceec5f5eac3825..9c12b310199cce330977fddca6d93c9d767552f2#r590228515

sqrrm reviewed (approved):
utACK

ghubstan commented:
Closed with merge of next PR https://github.com/bisq-network/bisq/pull/5284 (without file conflicts).
